### PR TITLE
Added the Ability to "Open" the Fairing in the Editor to allow seeing and editing the payload inside

### PR DESCRIPTION
**Pull request by pap1723**

### CODE REVIEW

File: Source/ProceduralFairings/FairingBase.cs [hunk 1]
```diff
@@ -491,10 +505,13 @@ public void UpdateFairingSideDragCubes()
...
-                        sp.GetComponent<ProceduralFairingSide>() is ProceduralFairingSide sf &&
-                        sf.dragCubeUpdater is DragCubeUpdater)
+                        sp.GetComponent<ProceduralFairingSide>() is ProceduralFairingSide sf)
                     {
-                        sf.dragCubeUpdater.Update();
+                        sides.Add(sf);
+                        Debug.Log($"sides: {sides.ToString()}");
+                        
+                        if (sf.dragCubeUpdater is DragCubeUpdater)
+                            sf.dragCubeUpdater.Update();
```
Comment (DRVeyl) [inline Source/ProceduralFairings/FairingBase.cs:514]: Don't do this here.  It doesn't clear the sides list, so every time this gets called it keeps appending all the sides that are attached.  But you really don't need to discover the sides here, anyway.  I'd undo this piece, remove the List<ProceduralFairingSide> declaration, and implement below.

File: Source/ProceduralFairings/FairingBase.cs [hunk 2]
```diff
@@ -714,6 +731,19 @@ public void ShowHideInterstageNodes()
...
+            if (openFairing && HighLogic.LoadedSceneIsEditor)
+            {
+                editorOpenOffset.x = OffsetValueX;
+                sides.ForEach(side => side.SetOffset(side.meshPos + editorOpenOffset));
+            }
+            else
+            {
+                sides.ForEach(side => side.SetOffset(side.meshPos));
+            }
+        }
```
Comment (DRVeyl) [inline Source/ProceduralFairings/FairingBase.cs:745]: Either go collect the list of sides in here, or write a utility method for it.  Don't run on a global list that you have to maintain.  There's several ways to write this, but a 1-line LINQ one is:
```
var sides = part.FindAttachNodes("connect").Where(x => x.attachedPart is Part sp && sp.GetComponent<ProceduralFairingSide>() is ProceduralFairingSide).ToList();
```
Also see comment above about not needing to set `editorOpenOffset.x` directly.

File: Source/ProceduralFairings/FairingBase.cs [hunk 3]
```diff
@@ -414,6 +424,10 @@ void OnChangeAutoshapeUI(BaseField bf, object obj)
         void OnChangeShapeUI(BaseField bf, object obj) => UpdateShape(true);
         public void OnNodeVisibilityChanged(BaseField f, object obj) => ShowHideInterstageNodes();
 
+        public void OnToggleOpen(BaseField field, object oldValue) => UpdateOpen();
+
+        private float OffsetValueX => Math.Max(minOffset, Math.Max(baseSize, topSize));
```
Comment (DRVeyl) [inline Source/ProceduralFairings/FairingBase.cs:429]: Move this up to line ~127 with the things that will use it?

File: Source/ProceduralFairings/FairingBase.cs [hunk 4]
```diff
@@ -117,6 +121,9 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         readonly List<ConfigurableJoint> joints = new List<ConfigurableJoint>();
         public DragCubeUpdater dragCubeUpdater;
         public ModuleDecouple Decoupler;
+        private readonly List<ProceduralFairingSide> sides = new List<ProceduralFairingSide>();
+        public Vector3 editorOpenOffset = new Vector3(3, 0, 0);
+        private float minOffset = 2f;
```
Comment (DRVeyl) [inline Source/ProceduralFairings/FairingBase.cs:126]: Trick:
`public Vector3 EditorOpenOffset => new Vector3(OffsetValueX, 0, 0);`

Then you won't have to mess with trying to set the x value manually.  Whenever you ask for the EditorOpenOffset Vector3, you will get the appropriate one.

Declare `minOffset` as `const` ie `private const float minOffset = 2`

File: Source/ProceduralFairings/FairingBase.cs [hunk 5]
```diff
@@ -346,6 +353,9 @@ void SetUIChangedCallBacks()
 
             Fields[nameof(showInterstageNodes)].uiControlEditor.onFieldChanged += OnNodeVisibilityChanged;
             Fields[nameof(showInterstageNodes)].uiControlEditor.onSymmetryFieldChanged += OnNodeVisibilityChanged;
+            
+            Fields[nameof(openFairing)].uiControlEditor.onFieldChanged = OnToggleOpen;
+            Fields[nameof(openFairing)].uiControlEditor.onSymmetryFieldChanged = OnToggleOpen;
```
Comment (DRVeyl) [inline Source/ProceduralFairings/FairingBase.cs:358]: I guess these should be `+=` if everything else is.  (Although.. in OnStart... practically every other mod I've been involved in will do `=` during initialization.  So... well, `+=` is probably better anyway.)

**Review by DRVeyl — CHANGES_REQUESTED** [review]

**Comment by pap1723** [comment]
This Branch has been corrupted and combined, moved to a new branch.

**Pull request closed without merging**